common/web/types/src/ldml-keyboard/pattern-parser.ts

```diff
@@ -275,48 +275,45 @@ export class MarkerParser {
     }
     /** output string */
     let out = '';
-    /** for checking: the total number of markers expected */
-    const max_markers = map.length;
+    /** for checking: the total number of markers expected, skipping end markers */
+    const max_markers = map.filter(({ end }) => !end).length;
     /** for checking: the number of markers we've written */
     let written_markers = 0;
     /** we are going to mutate the map, so copy it */
     const map2 : MarkerMap = [...map]; // make a copy
     // First, add back all 'MARKER_BEFORE_EOT' markers
     while (map2.length && map2[map2.length - 1].ch === MARKER_BEFORE_EOT) {
       // remove from list
-      const { marker } = map2.pop();
-      out = MarkerParser.prepend_marker(out, marker, forMatch);
-      written_markers++;
+      const { marker, end } = map2.pop();
+      if (!end) {
+        out = MarkerParser.prepend_marker(out, marker, forMatch);
+        written_markers++;
+      }
     }
     // Then, take each codepoint (from back to front)
     for (let p of [...s].reverse()) {
       // reverse order code units, prepend to out
       out = p + out;
 
-      // 1. we write any markers which match the output string in order,
-      // and pop them off the list as dones
-      while(map2.length > 0 && map2[map2.length-1].ch === p) {
-        const { marker } = map2.pop();
-        if (marker !== constants.marker_no_index) { // if not already written
-          out = MarkerParser.prepend_marker(out, marker, forMatch);
-          written_markers++;
-          // no need to update .marker here, we're about to pop it
-        }
-      }
```

Comment:
I took out this 'quick check' part.  Instead, we attempt to pop off (shorten) the list where possible.